internal/tofu/eval_context_builtin.go

```diff
@@ -69,6 +69,8 @@ type BuiltinEvalContext struct {
 	ProviderLock          *sync.Mutex
 	ProvisionerCache      map[string]provisioners.Interface
 	ProvisionerLock       *sync.Mutex
+	FunctionCache         *ProviderFunctions
+	FunctionLock          sync.Mutex
 	ChangesValue          *plans.ChangesSync
 	StateValue            *states.SyncState
 	ChecksValue           *checks.State
```

Comment:
It's currently inverted in this struct.  Your example conflates provider and provisioner.

```
ProvisionerCache ...
ProvisionerLock ...

FunctionCache ...
FunctionLock ...
```

Reply:
Yes, so that's why we would want to reorganize the entire field list instead of only the Function Cache and Lock. It is just a bit confusing at first glance.

Reply:
Ok, I'd rather not refactor that struct ordering in this PR.  There's a few places where we could change this in a followup PR if you'd like.

Reply:
I'm in favour of doing this (in a followup PR).

Reply:
Can we raise an issue for this so we don't forget then please?